internal/srtp/session_srtcp.go

```diff
@@ -0,0 +1,137 @@
+package srtp
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net"
+
+	"github.com/pions/webrtc/pkg/rtcp"
+)
+
+// SessionSRTCP implements io.ReadWriteCloser and provides a bi-directional SRTCP session
+// SRTCP itself does not have a design like this, but it is common in most applications
+// for local/remote to each have their own keying material. This provides those patterns
+// instead of making everyone re-implement
+type SessionSRTCP struct {
+	session
+	writeStream *WriteStreamSRTCP
+}
+
+// CreateSessionSRTCP creates a new SessionSRTCP
+func CreateSessionSRTCP() *SessionSRTCP {
+	s := &SessionSRTCP{}
+	s.writeStream = &WriteStreamSRTCP{s}
+	s.session.initalize()
+	return s
+}
+
+// Start initializes any crypto context and allows reading/writing to begin
+func (s *SessionSRTCP) Start(localMasterKey, localMasterSalt, remoteMasterKey, remoteMasterSalt []byte, profile ProtectionProfile, nextConn net.Conn) error {
+	s.session.nextConn = nextConn
+	return s.session.start(localMasterKey, localMasterSalt, remoteMasterKey, remoteMasterSalt, profile, s)
+}
+
+// OpenWriteStream returns the global write stream for the Session
+func (s *SessionSRTCP) OpenWriteStream() (*WriteStreamSRTCP, error) {
+	return s.writeStream, nil
+}
+
+// OpenReadStream opens a read stream for the given SSRC, it can be used
+// if you want a certain SSRC, but don't want to wait for AcceptStream
+func (s *SessionSRTCP) OpenReadStream(SSRC uint32) (*ReadStreamSRTCP, error) {
+	r, _ := s.session.getOrCreateReadStream(SSRC, s, &ReadStreamSRTCP{})
+
+	if readStream, ok := r.(*ReadStreamSRTCP); ok {
+		return readStream, nil
+	}
+	return nil, fmt.Errorf("Failed to open ReadStreamSRCTP, type assertion failed")
+}
+
+// AcceptStream returns a stream to handle RTCP for a single SSRC
+func (s *SessionSRTCP) AcceptStream() (*ReadStreamSRTCP, uint32, error) {
+	stream, ok := <-s.newStream
+	if !ok {
+		return nil, 0, fmt.Errorf("SessionSRTCP has been closed")
+	}
+
+	readStream, ok := stream.(*ReadStreamSRTCP)
+	if !ok {
+		return nil, 0, fmt.Errorf("newStream was found, but failed type assertion")
+	}
+
+	return readStream, stream.GetSSRC(), nil
+}
+
+// Close ends the session
+func (s *SessionSRTCP) Close() error {
+	return s.session.close()
+}
+
+// Private
+
+func (s *SessionSRTCP) write(buf []byte) (int, error) {
+	if _, ok := <-s.session.started; ok {
+		return 0, fmt.Errorf("started channel used incorrectly, should only be closed")
+	}
+
+	s.session.localContextMutex.Lock()
+	defer s.session.localContextMutex.Unlock()
+
+	encrypted, err := s.localContext.EncryptRTCP(nil, buf, nil)
+	if err != nil {
+		return 0, err
+	}
+	return s.session.nextConn.Write(encrypted)
+}
+
+func (s *SessionSRTCP) decrypt(buf []byte) error {
+	decrypted, err := s.remoteContext.DecryptRTCP(nil, buf, nil)
```

Comment:
RTCP packets usually contain multiple SSRces (so we can't decrypt in place I don't think)